feat(oracledb): Add support for Oracle DB instrumentation

sudarshan12s opened the pull request:
Add Oracledb instrumentation

<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/guides/contributor#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?
Adding instrumentation support for [node-oracledb ](https://github.com/oracle/node-oracledb)applications connecting to Oracle database. 


## Short description of the changes
From [node-oracledb ](https://github.com/oracle/node-oracledb) version 6.7 onwards,  oracledb module exports a class, 
`TraceHandlerBase` which simulates interface kind of functionality with abstract methods to be overwritten by the derived class.

The current OT module derives this class `TraceHandlerBase` and implements abstract methods; It uses `setTraceInstance` method of `oracledb` module to register this new derived class `OracleTelemetryTraceHandler`
These methods are invoked by `oracledb` module providing the  `traceContext` which is used to dump the data in to the span after converting to standard span attributes.

Following are the abstract methods implemented:

- onEnterFn -> It gets invoked by `oracledb` module when application invokes `getConnection` to create a connection to 
database or `execute` method to run a sql. It is invoked before invoking the public API calls made by application.
- onExitFn -> It gets invoked after the public API call is completed with success/failure.
- onBeginRoundTrip -> It gets invoked before round trip to database which was triggered as part of public API.
- onEndRoundTrip -> It gets invoked after round trip to database completes which was triggered as part of public API..

raphael-theriault-swi reviewed (changes requested):
Could you rename the new types to TitleCase to follow the rest of the code ?

sudarshan12s commented:
> Could you rename the new types to TitleCase to follow the rest of the code ?

Yes renamed to TitleCase for types and internal-types exported.

sudarshan12s commented:
The earlier triggered CI failure seems to be this. It looks related to some package version dependency ,` package-lock.json` updates and not the changes in PR.   Please suggest if there are some changes needed from my side.  
```

 Exception during run: /home/runner/work/opentelemetry-js-contrib/opentelemetry-js-contrib/node_modules/@aws-sdk/client-sqs/dist-cjs/index.js:185
  static {
         ^

SyntaxError: Unexpected token '{'
```

sudarshan12s commented:
Please help me with steps for updating` package-lock.json` . This PR lock file  ` package-lock.json` differs from main and i am not able to keep it in sync easily..

sudarshan12s commented:
> Please help me with steps for updating` package-lock.json` . This PR lock file ` package-lock.json` differs from main and i am not able to keep it in sync easily..

I have fixed this discrepency.

JamieDanielson commented:
@sudarshan12s as we discussed during SIG meeting yesterday, please review the [requirements here for component ownership](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/CONTRIBUTING.md#component-ownership), as well as our [preferred approach for long-term maintenance of instrumentation libraries](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/CONTRIBUTING.md#new-instrumentation).

marcalff commented:
> @sudarshan12s as we discussed during SIG meeting yesterday, please review the [requirements here for component ownership](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/CONTRIBUTING.md#component-ownership), as well as our [preferred approach for long-term maintenance of instrumentation libraries](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/CONTRIBUTING.md#new-instrumentation).

To @JamieDanielson 
Cc @open-telemetry/javascript-maintainers

Hi Jamie, All,

To become a member of CNCF, @sudarshan12s will need two sponsors for his application.

I am myself an Oracle employee, one of the [maintainers](https://github.com/open-telemetry/opentelemetry-cpp?tab=readme-ov-file#contributing) of the opentelemetry-cpp repository, and can sponsor Sudarshan.

Could you, or someone in javascript-maintainers, be the second sponsor ?
I think it makes the most sense, since Sudarshan contribution(s) are in the js-contrib space.

Regards,
-- Marc

sudarshan12s commented:
> Hi @sudarshan12s. I was absent during last week's SIG meeting so I'm not sure if there was an agreement made to not apply the current guidelines, but there are rules and guidelines for new instrumentations. These were introduced since many instrumentations were abandoned in this repository and we would like to avoid this in the future as it diverts resources from development of OTel core components.
> 
> **Before opening a PR,** the following questions should be answered, see [Jamie's link](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/CONTRIBUTING.md#new-instrumentation):
> 
> * who's going to maintain it (2 people minimum, both must be members of the OTel org prior to opening the PR)
> * why not host it in a separate repo? (hosting it here has maintenance implications for the repo, the less we have to host here, the better the quality of other components like the API and SDK is going to be as we get to spend more time on that)
> 
> This is done in the form of an [instrumentation request](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/new?template=instrumentation_request.md).

Thanks @pichlermarc , Yes we were informed on the guidelines/procedure to be followed in the last SIG. I have updated the instrumentation request [here](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/2703) with more details. Let me know if this looks fine. For becoming members of OTel org, @marcalff  has agreed for first sponsor. Can someone from javascript-maintainers be the second sponsor.

blumamir commented on plugins/node/opentelemetry-instrumentation-oracledb/src/constants.ts:
```diff
@@ -0,0 +1,57 @@
+// Contains span names produced by instrumentation
+// It lists the RPC names (suffix with _MSG like EXECUTE_MSG) and
+// exported oracledb functions (like EXECUTE).
+export enum SpanNames {
```
please see the guidelines for span names [here](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-spans.md#name) and review if we can improve the span name. 
I understand not all the info is available.

blumamir commented on plugins/node/opentelemetry-instrumentation-oracledb/src/utils.ts:
```diff
@@ -0,0 +1,325 @@
+  ATTR_DB_NAMESPACE,
+  ATTR_DB_OPERATION_NAME,
+} from './semconv';
+import * as oracledbTypes from 'oracledb';
```
```suggestion
import type * as oracledbTypes from 'oracledb';
```

sudarshan12s replied:
The type script [definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/oracledb) still does not have new exported class [`TraceHandlerBase`](https://github.com/oracle/node-oracledb/blob/main/lib/traceHandler.js)  , Hence importing both types and values. 


Should I keep it for now and remove as types definitions get updated?

sudarshan12s replied:
The TraceHandlerBase is modelled as abstract class, hence I need to extend and not implement. I made changes to use dynamic require with a function, Please see and let me know if this is fine

blumamir commented on plugins/node/opentelemetry-instrumentation-oracledb/src/utils.ts:
```diff
@@ -0,0 +1,325 @@
+ */
+export class OracleTelemetryTraceHandler extends newmoduleExports.traceHandler
+  .TraceHandlerBase {
+  private _tracer: Tracer;
```
the tracer should be updated when `setTracerProvider` is called on the instrumentation class.

sudarshan12s replied:
I am passing the function (`() => this.tracer` ) to OracleTelemetryTraceHandler constructor instead. Just to avoid override of `setTracerProvider` . looks fine?

blumamir commented on plugins/node/opentelemetry-instrumentation-oracledb/src/internal-types.ts:
```diff
@@ -0,0 +1,46 @@
+  additionalConfig?: any; // custom key/values associated with a function.
+  fn: Function; // Replaced with bind function associating the active context.
+  args?: any[]; // input arguments passed to the exported function.
+  userContext: InstrumentationContext;
```
please add a short comment about this field, and how the driver uses it, so it's documented for future maintainers

blumamir commented on plugins/node/opentelemetry-instrumentation-oracledb/src/internal-types.ts:
```diff
@@ -0,0 +1,46 @@
+  span: api.Span;
+}
+
+// Captures the entire span data
```
you can add a short comment documenting that this interface correlates to a js object that is exported from the instrumented package. preferably with a link

sudarshan12s replied:
Added following:
```

// This corresponds to js object filled by the 'oracledb' module
// See: https://github.com/oracle/node-oracledb/blob/main/lib/traceHandler.js
```

blumamir commented on plugins/node/opentelemetry-instrumentation-oracledb/src/constants.ts:
```diff
@@ -0,0 +1,57 @@
+  ORACLE_IMPLICIT_RELEASE = 'db.oracle.implicit_release',
+}
+
+// Contains span names produced by instrumentation
```
These span names actually correlates to the values being produced by the instrumented package itself, let's add it as a comment so people will know it's not arbitrary and will break if changed

blumamir commented on plugins/node/opentelemetry-instrumentation-oracledb/src/utils.ts:
```diff
@@ -0,0 +1,325 @@
+  // completes.
+  onExitFn(traceContext: TraceSpanData) {
+    if (
+      this._shouldSkipInstrumentation() ||
+      !traceContext.userContext ||
+      !traceContext.userContext.span
```
```suggestion
      !traceContext.userContext?.span
```

blumamir commented on plugins/node/opentelemetry-instrumentation-oracledb/src/utils.ts:
```diff
@@ -0,0 +1,325 @@
+      );
+    }
+    if (traceContext.error) {
+      span.setStatus({
```
consider:

```
  span.recordException(traceContext.error);
```

which will log a span event for this error

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/constants.ts:
```diff
@@ -0,0 +1,57 @@
+ * Copyright (c) 2024, Oracle and/or its affiliates.
+ * */
+
+// Oracle specific attributes not covered by semantic conventions
```
I am working on the [PR](https://github.com/open-telemetry/semantic-conventions/pull/1911) for oracledb semconv.

sudarshan12s replied:
Link is also added in the code for the same.

blumamir commented:
> For becoming members of OTel org, @marcalff has agreed for first sponsor. Can someone from javascript-maintainers be the second sponsor.

Once this PR is finalized and ready to merge, I can be your second sponsor for the organization membership

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/types.ts:
```diff
@@ -0,0 +1,92 @@
+export interface OracleInstrumentationConfig extends InstrumentationConfig {
+  /**
+   * If true, an attribute containing the execute method
+   * bind values will be attached the spans generated.
```
changed as following:
```
If true, an attribute containing the execute method
   bind values will be attached the spans generated.
  It can potentially record PII data and should be used with caution.
```

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/utils.ts:
Renamed to `OracleTelemetryTraceHandler.ts`

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/types.ts:
```diff
@@ -0,0 +1,92 @@
+  enhancedDatabaseReporting?: boolean;
+
+  /**
+   * If true, db.statement will have sql in the spans generated.
```
Added the following:
```

If true, db.statement will have sql which could potentially contain
    sensitive unparameterized data in the spans generated.
```

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/instrumentation.ts:
```diff
@@ -0,0 +1,68 @@
+      'oracledb',
+      ['>= 6.7 < 7'],
+      (moduleExports: typeof oracleDBTypes) => {
+        const newmoduleExports: any = moduleExports;
```
Yes removed `newmoduleExports` variable.

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/instrumentation.ts:
```diff
@@ -0,0 +1,68 @@
+
+        // Register the instance with oracledb.
+        newmoduleExports.traceHandler.setTraceInstance(obj);
+        this._tmHandler = obj;
```
I am using non null check to see if its patched already, then unregister and register again. Please add if i missed anything.

`if (this._tmHandler) {`

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/types.ts:
```diff
@@ -0,0 +1,92 @@
+
+export interface OracleRequestHookInformation {
+  inputArgs: any;
+  connection: SpanConnectionConfig | undefined;
```
removed `undefined`.

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/OracleTelemetryTraceHandler.ts:
```diff
@@ -0,0 +1,439 @@
+import type { traceHandler } from 'oracledb'; // Import only for type checking
+
+// It returns the TraceHandlerBase class, if oracledb module is available.
+function getTraceHandlerBaseClass(): TraceHandlerBaseConstructor | null {
```
For the issue with requiring oracledb to be present in node_modules with this line:
`import * as oracledbTypes from 'oracledb';`

I made changes to import types instead:
`import type { traceHandler } from 'oracledb';`

Now since `oracledb` module exposes `TraceHandlerBase` as abstract class where some default implementations (`enable`) are available, I can not do `implements TraceHandlerBase`.
Added a function, getTraceHandlerBaseClass which gives the constructor for TraceHandlerBase dynamically during patch avoiding global `require("oracledb")` in generated js files. 

The types added inside this file will be removed once this [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/72060) and release including this change is done.

Please confirm if i am missed anything or any suggestions.


Thanks for pointing it out with details.

raphael-theriault-swi replied:
Note that the instance of `oracledb` that's being `require`d by the instrumentation package here might differ from the one that is being patched. Given you have access to the exports from the patch method it would be a better idea to pass them here as a parameter and get the base class from the instance that's being instrumented.

sudarshan12s replied:
got it. I made the changes accordingly.

raphael-theriault-swi replied:
Thank you !

sudarshan12s commented:
> Did another round of review and added few more comments.
> 
> Currently marking as "RequestChanges" due to the following 2 issues:
> 
> 1. the copyright to oracle should be addressed (not sure in what way) @open-telemetry/javascript-maintainers
> 2. the instrumentation must not import from the instrumented package.
> 
> Thanks again

I have addressed the second point by including types , Please see.

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/utils.ts:
```diff
@@ -0,0 +1,330 @@
+        if (this._instrumentConfig.enhancedDatabaseReporting && !roundTrip) {
+          const values = this._getValues(callConfig.values);
+          if (values) {
+            span.setAttribute(AttributeNames.ORACLE_BIND_VALUES, values);
```
I will try. But since it was more for debugging and its a sensitive information, I think its less priority.

sudarshan12s replied:
I have added it now along with other feedback from semantic conventions PR.

dyladan commented:
https://github.com/cncf/foundation/blob/main/copyright-notices.md#what-if-i-want-my-copyright-notice-included

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/OracleTelemetryTraceHandler.ts:
```diff
@@ -0,0 +1,372 @@
+  obj: typeof oracleDBTypes
+): any {
+  const traceHandlerBase = getTraceHandlerBaseClass(obj);
+  if (traceHandlerBase) {
```
Done.

sudarshan12s commented:
Hey @opentelemetry/javascript-approvers, I've incorporated the feedback received so far. If anyone else would like to take a look, I'd appreciate additional inputs!

Thanks.

maryliag commented on plugins/node/opentelemetry-instrumentation-oracledb/src/constants.ts:
```diff
@@ -0,0 +1,59 @@
+
+// Oracle specific attributes not covered by standard semantic conventions.
+// See: https://github.com/open-telemetry/semantic-conventions/pull/1911
+export enum AttributeNames {
```
I don't think this PR should be merged until the [Semantic Conventions PR ](https://github.com/open-telemetry/semantic-conventions/pull/1911) for it is merged, since the definition can change there and you don't want to have something merged that is already out of date even before it gets released. And on that topic, why these attributes were not added to the semantic convention? 
I imagine the goal is to have a similar instrumentation for oracle in the other SDKs, but how will Java or .NET know what names to use if they're not defined in the Semantic convention. They will have no idea that it was already implemented in JS and they should follow here. This is why the Semantic Conventions are so important. So please add a complete definition there with all attributes you're using in this instrumentation 😄 

Also, as I mentioned in the SIG, really try to use already existing name attributes, for example, we have several [attributes for pool](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-metrics.md#connection-pools) already, so there is no need to create specific with oracle in the name. 
You should check if we don't have for the others.

sudarshan12s replied:
Thanks, @maryliag! I've started making the changes based on the inputs suggested in the semantic conventions PR.

I’ll use the existing attributes and retain only one entry in `constants.ts`, as it is specific to Oracle [DRCP](https://node-oracledb.readthedocs.io/en/latest/user_guide/connection_handling.html#implicitpool):
`ORACLE_IMPLICIT_RELEASE = 'oracle.db.implicit_release',` 

For pool attributes, I plan to map them as follows:
```
oracledb.pool.max -> db.client.connection.max  
oracledb.pool.min -> db.client.connection.min  
```
I also requested clarification on whether `db.client.connection.max` and `db.client.connection.min` can be used in DB spans when the `oracledb.createPool` API is called, in addition to being included in metrics. They will not be dumped in subsequent DB queries.

For now, I’ve removed `oracledb.pool_incr` (I have not seen in existing attributes) but can add it back to the semantic conventions and instrumentation once I receive clarification on the above.

sudarshan12s replied:
I have incorporated the changes based on the semantic conventions PR feedback received so far. The attribute `oracle.db.implicit_release` is still under review.

sudarshan12s replied:
The semantic conventions PR is merged now. I made changes to incorporate them. Please see. 

Following is the example of span attributes aligning to conventions.

If a sql query, `'select :name from dual'` is executed with name being the placeholder , following attributes would appear.


```
spanName-> oracledb.Connection.execute:SELECT FREE|FREEPDB1|freepdb1 (It has Fnname:operationName db.namespace)
db.operation.name ='SELECT'
db.operation.parameter.name ='0'
db.query.text = 'select :name from dual'
db.system.name ='oracle.db'
db.namespace = 'FREE|FREEPDB1|freepdb1'. (It has instanceName | pdbname | serviceName)
```

maryliag reviewed (commented):
Thank you for your work on this!
I added a few comments

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/constants.ts:
```diff
@@ -0,0 +1,47 @@
+}
+
+/*
+ * The semantic conventions defined DBSYSTEMVALUES_ORACLE as oracle, hence
```
This was sourced from [here](https://github.com/open-telemetry/semantic-conventions/blob/dd6a536db4a2c7d351f0cae72bd253f2955da25b/model/database/registry.yaml#L301) and also reflects a suggestion to follow that format. I think the intent is to avoid including vendor or company names, and instead have the engine or database.

maryliag replied:
got it, thanks

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/types.ts:
```diff
@@ -0,0 +1,94 @@
+   *
+   * @default false
+   */
+  dbStatementDump?: boolean;
```
I have updated documentation with more details. `dbStatementDump` only dumps sql stmt but never bind values. But `enhancedDatabaseReporting` dumps sql stmt and bind values.. This was done to enable basic debugging of sql stmts .. Looks fine?

maryliag replied:
thanks for the explanation!

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/test/oracle.test.ts:
```diff
@@ -0,0 +1,1680 @@
+
+import {
+  ATTR_SERVER_ADDRESS,
+  SEMATTRS_NET_TRANSPORT,
```
Yes replaced with ATTR_*

maryliag replied:
I'm still seeing a lot of `SEMATTRS_` here

sudarshan12s replied:
I had missed the test files earlier. added now—thanks.

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/test/oracle.test.ts:
```diff
@@ -0,0 +1,1680 @@
+// It verifies the spans, its attributes and the parent child relationship.
+function verifySpans(
+  parentSpan: Span | null,
+  attributesList: Attributes[] = [executeAttributes, executeAttributes],
```
For executing a query, typically we have one span for public API and a child span doing a round trip. 

Both the spans have same attributes dumped, hence we are passing same attributes for both public API and round trip span..

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/OracleTelemetryTraceHandler.ts:
```diff
@@ -0,0 +1,434 @@
+          span.recordException(traceContext.error);
+          span.setStatus({
+            code: SpanStatusCode.ERROR,
+            message: traceContext.error.message,
```
Thanks. I shall confirm on this.

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/src/OracleTelemetryTraceHandler.ts:
```diff
@@ -0,0 +1,434 @@
+        if (typeof value === 'object') {
+          return JSON.stringify(value);
+        }
+        // number, string, boolean,
```
Donw

sudarshan12s commented on plugins/node/opentelemetry-instrumentation-oracledb/README.md:
```diff
@@ -0,0 +1,73 @@
+| Options | Type | Default | Description |
+| ------- | ---- | --------| ----------- |
+| `enhancedDatabaseReporting` | `boolean` | `false` | If true, additional details about the sql statement's bind values will be attached to the spans generated to represent the database operations |
+| `dbStatementDump` | `boolean` | `false` | If true, ``db.statement`` will contain the sql string in the spans generated |
```
Corrected.

sudarshan12s commented:
> Thank you for your work on this! I added a few comments

I have incorporated the comments..  Please see.

maryliag commented on plugins/node/opentelemetry-instrumentation-oracledb/src/semconv.ts:
```diff
@@ -0,0 +1,113 @@
+ * The connection string used to connect to the database. It is recommended to remove embedded credentials.
+ *
+ */
+export const ATTR_DB_CONNECTION_STRING = 'db.connection_string';
```
this attribute is deprecated
``@deprecated Replaced by `server.address` and `server.port`.``

sudarshan12s replied:
Can I keep this attribute. 

The reason is : 

The [node-oracledb](https://github.com/oracle/node-oracledb) driver supports a thick mode, in which the address and port are not directly available. While it's technically possible to extract them by parsing the connection string, this approach is avoided. Moreover, if the connection string is an [alias](https://node-oracledb.readthedocs.io/en/latest/user_guide/connection_handling.html#tns-aliases-for-connection-strings) referencing an external configuration file (such as a tnsnames.ora entry), the address and port may not be accessible at all . It is planned to expose these attributes in thick mode from node-oracledb driver itself but currently they are not available.

This limitation does not apply to thin mode in [node-oracledb](https://github.com/oracle/node-oracledb), as it provides the parsed address, port, and protocol directly.

sudarshan12s replied:
I have removed the `ATTR_DB_CONNECTION_STRING` attribute as suggested in the SIG.

maryliag reviewed (commented):
Thank you for addressing all comments. It's almost there :) 
I added a few more comments, all easy to fix!

maryliag commented on plugins/node/opentelemetry-instrumentation-oracledb/test/oracle.test.ts:
```diff
@@ -0,0 +1,1675 @@
+const CONFIG = {
+  user: process.env.ORACLE_USER || 'demo',
+  password: process.env.ORACLE_PASSWORD || 'demo',
+  connectString: process.env.ORACLE_CONNECTSTRING || 'localhost:1521/freepdb1',
```
```suggestion
```

sudarshan12s replied:
> 

This is a configuration attribute that contains information such as `host`, `port`, and `servicename`. It's required to establish a database connection and run the tests. However, it is never validated within the span, as it does not appear there.

maryliag reviewed (approved):
Great work on this!

sharadraju commented:
Thanks @maryliag and @blumamir  for your review and thanks @marcalff for the sponsor support!
@sudarshan12s Nice work!

sudarshan12s commented:
> Great work on this!

Thank you so much...

sudarshan12s commented:
It looks like the failed test may be flaky (intermittently asserting different connection counts). Would it be possible to either ignore this failure for this PR, or re-run the CI?

```
Expected values to be strictly equal:
2 !== 1
at test/mongodb-v4-v5-v6.metrics.test.ts:127:12
```

Thanks for your time!`

sudarshan12s commented:
> It looks like the failed test may be flaky (intermittently asserting different connection counts). Would it be possible to either ignore this failure for this PR, or re-run the CI?
> 
> ```
> 
> MongoDBInstrumentation-Metrics
>        Should add connection usage metrics:
> 
>       AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
> 
> 2 !== 1
> 
>       + expected - actual
> 
>       -2
>       +1
>       
>       at Context.<anonymous> (test/mongodb-v4-v5-v6.metrics.test.ts:127:12)
> ```
> 
> Thanks for your time!

@onurtemizkan, I believe this is a flaky error and can be safely ignored.

sharadraju commented:
Thanks @blumamir for merging this PR.
